Target .NET 5

UselessToucan opened the pull request:
- [x] Needs testing on mobile platforms (iOS in particular)

UselessToucan committed:
Target net5.0 instead of netcoreapp3

UselessToucan committed:
Update NuGet packages

UselessToucan committed:
Fix FileNotFoundException on startup

UselessToucan committed:
Disable CA1416

UselessToucan committed:
Resolve CA1416 properly using new API

UselessToucan committed:
Workaround FileNotFoundException in a test projects

peppy commented on osu.Game/osu.Game.csproj:
```diff
@@ -18,11 +18,13 @@
     <PackageReference Include="Humanizer" Version="2.8.26" />
-    <PackageReference Include="Microsoft.AspNetCore.SignalR.Client" Version="3.1.9" />
-    <PackageReference Include="Microsoft.AspNetCore.SignalR.Protocols.NewtonsoftJson" Version="3.1.9" />
+    <PackageReference Include="Microsoft.AspNetCore.SignalR.Client" Version="5.0.0" />
```
will need re-testing of mobile platforms as we have them pinned to a much earlier version for compatibility reasons.

peppy replied:
These changes were reverted, no longer an issue.

UselessToucan committed:
Fix InspectCode warnings

huoyaoyuan commented:
I'd suggest to separate this into multiple parts. Updating SDK and entry executable to .NET 5 can unblock SDK feature opt-in.

peppy committed:
Merge branch 'master' into net5.0

peppy committed:
Revert polling changes to fix participant list display

It turns out this polling was necessary to get extra data that isn't
included in the main listing request. It was removed deemed useless, and
in order to fix the order of rooms changing when selecting a room.
Weirdly, I can't reproduce this happening any more, and on close
inspection of the code can't see how it could happen in the first place.

For now, let's revert this change and iterate from there, if/when the
same issue arises again.

I've discussed avoiding this second poll by potentially including more
data (just `user_id`s?) in the main listing request, but not 100% sure
on this - even if the returned data is minimal it's an extra join
server-side, which could cause performance issues for large numbers of
rooms.

peppy committed:
Fix unstable multiplayer room ordering when selection is made

peppy committed:
Fix deadlock scenario when calculating fallback difficulty

The previous code would run a calcaulation for the beatmap's own ruleset
if the current one failed. While this does make sense, with the current
way we use this component (and the implementation flow) it is quite unsafe.

The to the call on `.Result` in the `catch` block, this would 100%
deadlock due to the thread concurrency of the `ThreadedTaskScheduler`
being 1. Even if the nested run could be run inline (it should be), the
task scheduler won't even get to the point of checking whether this is
feasible due to it being saturated by the already running task.

I'm not sure if we still need this fallback lookup logic. After removing
it, it's feasible that 0 stars will be returned during the scenario that
previously caused a deadlock, but I don't necessarily think this is
incorrect. There may be another reason for this needing to exist which
I'm not aware of (diffcalc?) but if that's the case we may want to move
the try-catch handling to the point of usage.

To reproduce the deadlock scenario with 100% success (the repro
instructions in the linked issue aren't that simple and require some
patience and good timing), the main portion of the lookup can be changed
to randomly trigger a nested lookup:

```
if (RNG.NextSingle() > 0.5f)
    return GetAsync(new
DifficultyCacheLookup(key.Beatmap, key.Beatmap.Ruleset,
key.OrderedMods)).Result;
else
    return new StarDifficulty(attributes);
```

After switching beatmap once or twice, pausing debug and viewing the
state of threads should show exactly what is going on.

UselessToucan committed:
Merge branch 'master' into net5.0

peppy committed:
Update framework

peppy committed:
Update KeyBinding usages in line with interface changes

peppy committed:
Update some missed usages of KeyBindingContainer in tests

peppy committed:
Update framework (again) for native libs fix

peppy committed:
Remove unnecessary workaround (mentioned package is pinned by SignalR to a working version)

peppy committed:
Remove global.json

peppy committed:
Restore previous exception handling flow for stable path lookup

peppy committed:
Remove assert messages

peppy committed:
Merge branch 'master' into net5.0

bdach commented:
Tested android and windows, from basic checks it seems to work fine. I suppose time to get this in and see if anything falls out.

bdach committed:
Merge branch 'master' into net5.0
